File: ts/env.ts
import { Data } from "./types";

export class Env {
  outer: Env | null;
  data: { [K: string]: Data };
  constructor(outer: Env | null = null) {
    this.outer = outer;
    this.data = {};
  }

  set(key: string, value: Data) {
    this.data[key] = value;
  }

  find(key: string): Env {
    if (this.data[key] !== undefined) {
      return this;
    }
    if (this.outer !== null) {
      return this.outer.find(key);
    }
    throw new Error("not found : " + key);
  }

  get(key: string) {
    return this.find(key).data[key];
  }
}

File: impls/python/env.py
from reader import Val

class Env:
    def __init__(self, outer = None):
        self.data = {}
        self.outer = outer

    def set(self, key, value):
        if value.type != "error":
                self.data[key] = value

    def get(self, key):
        if key in self.data:
            return self.data[key]
        if self.outer != None:
            return self.outer.get(key)
        raise Exception(Val("string", "'" + key + "' not found"))
